src/content/docs/en/recipes/i18n.mdx

```diff
@@ -0,0 +1,327 @@
+          const pages = await getCollection('blog')
+
+          const paths = pages.map(page => {
+            const [lang, ...slug] = page.slug.split('/');
+            return { params: { lang, slug: slug.join('/') || undefined }, props: page }
+          })
+
+          return paths;
+        }
+
+        const { lang, slug } = Astro.params;
+        const page = Astro.props;
+        const formattedDate = page.data.date.toLocaleString(lang);
+
+        const { Content } = await page.render();
+        ---
+        <h1>{page.data.title}</h1>
+        <p>by {page.data.author} • {formattedDate}</p>
+        <Content/>
+        ```
+      </Fragment>
+
+      <Fragment slot="ssr">
+        In [SSR mode](/en/guides/server-side-rendering/), fetch the requested entry directly:
+
+        ```astro
+        //src/pages/[lang]/blog/[...slug].astro
+        ---
+        import { getEntryBySlug } from 'astro:content';
+
+        const { lang, slug } = Astro.params;
+        const page = await getEntryBySlug('blog', `${lang}/${slug}`);
+
+        if (!page) {
+          return Astro.redirect('/404');
+        }
+
+        const formattedDate = page.data.date.toLocaleString(lang);
+        const { Content, headings } = await page.render();
+        ---
+        <h1>{page.data.title}</h1>
+        <p>by {page.data.author} • {formattedDate}</p>
+        <Content/>
+        ```
+      </Fragment>
+    </StaticSsrTabs>
+
+    📚 Read more about [dynamic routing](/en/core-concepts/routing/#dynamic-routes).
+
+    :::tip[Date formatting]
+    The example above uses the built-in [`toLocaleString()` date-formatting method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString) to create a human-readable string from our frontmatter date.
+    This ensures the date and time are formatted to match the user’s language.
+    :::
+
+### Translate UI strings
+
+Create dictionaries of terms to translate the labels for UI elements around your site. This allows your visitors to experience your site fully in their language.
+
+1. Create a `src/i18n/ui.ts` file to store your translation strings in:
+
+    ```ts
+    // src/i18n/ui.ts
+    export const languages = {
+      en: 'English',
+      fr: 'Français',
+    };
+
+    export const defaultLang = 'en';
+    
+    export const ui = {
+      en: {
+        'nav.home': 'Home',
+        'nav.about': 'About',
+        'nav.twitter': 'Twitter',
+      },
+      fr: {
+        'nav.home': 'Acceuil',
+        'nav.about': 'À propos',
+      },
+    } as const;
+    ```
+    
+2. Create a helper function to detect the page language based on the current URL and a helper to let us get translations strings for different parts of our UI in `src/i18n/utils.ts`:
+
+    ```js
+    // src/i18n/utils.ts
+    import { ui, defaultLang } from './ui';
+    
+    export function getLangFromUrl(url: URL) {
+      const [, lang] = url.pathname.split('/');
+      return lang || defaultLang;
```

Comment:
```suggestion
      return (lang as keyof typeof ui) || defaultLang;
```
Fix type mismatch in `Nav.astro` example where `getLangFromUrl()` returns type `string` but `useTranslations()` expects type `keyof typeof ui`

Reply:
Hmmm. In this case I think we might want to fix it in the `useTranslations()` function. It would be lying to say `getLangFromUrl` always returns a key of the UI dictionary. 🤔

Reply:
Actually, perhaps the best here _is_ to guarantee this is always a valid language in this function, but that requires an extra if statement. I’ll update the full function.